rclc_lifecycle/src/rclc_lifecycle/rclc_lifecycle.c

```diff
@@ -260,3 +272,160 @@ rclc_lifecycle_execute_callback(
 }
+
+rcl_ret_t
+rclc_lifecycle_add_get_state_service(
+  rclc_lifecycle_node_t * lifecycle_node,
+  rclc_executor_t * executor)
+{
+  rclc_lifecycle_service_context_t context;
+  context.lifecycle_node = lifecycle_node;
+
+  rcl_ret_t rc = rclc_executor_add_service_with_context(
+    executor,
+    &lifecycle_node->state_machine->com_interface.srv_get_state,
+    &lifecycle_node->gs_req,
+    &lifecycle_node->gs_res,
+    rclc_lifecycle_get_state_callback,
+    &context);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(main, rclc_executor_add_service_with_context);
+    return -1;
+  }
+
+  return rc;
+}
+
+void
+rclc_lifecycle_get_state_callback(
+  const void * req,
+  void * res,
+  void * context)
+{
+  RCL_UNUSED(req);
+  lifecycle_msgs__srv__GetState_Response * res_in =
+    (lifecycle_msgs__srv__GetState_Response *) res;
+  rclc_lifecycle_service_context_t * context_in =
+    (rclc_lifecycle_service_context_t *) context;
+
+  rcl_lifecycle_state_machine_t * sm =
+    context_in->lifecycle_node->state_machine;
+
+  res_in->current_state.id = sm->current_state->id;
+  bool success = rosidl_runtime_c__String__assign(
+    &res_in->current_state.label,
+    sm->current_state->label);
+  if (!success) {
+    PRINT_RCLC_ERROR(
+      rclc_lifecycle_get_state_callback,
+      rosidl_runtime_c__String__assign);
+  }
+}
+
+rcl_ret_t
+rclc_lifecycle_add_get_available_states_service(
+  rclc_lifecycle_node_t * lifecycle_node,
+  rclc_executor_t * executor)
+{
+  rclc_lifecycle_service_context_t context;
+  context.lifecycle_node = lifecycle_node;
+
+  rcl_ret_t rc = rclc_executor_add_service_with_context(
+    executor,
+    &lifecycle_node->state_machine->com_interface.srv_get_available_states,
+    &lifecycle_node->gas_req,
+    &lifecycle_node->gas_res,
+    rclc_lifecycle_get_available_states_callback,
+    &context);
+  if (rc != RCL_RET_OK) {
+    PRINT_RCLC_ERROR(main, rclc_executor_add_service_with_context);
+    return -1;
+  }
+
+  return rc;
+}
+
+void
+rclc_lifecycle_get_available_states_callback(
+  const void * req,
+  void * res,
+  void * context)
+{
+  RCL_UNUSED(req);
+  lifecycle_msgs__srv__GetAvailableStates_Response * res_in =
+    (lifecycle_msgs__srv__GetAvailableStates_Response *) res;
+  rclc_lifecycle_service_context_t * context_in =
+    (rclc_lifecycle_service_context_t *) context;
+
+  rcl_lifecycle_state_machine_t * sm =
+    context_in->lifecycle_node->state_machine;
+
+  lifecycle_msgs__srv__GetAvailableStates_Response__init(res_in);
+  lifecycle_msgs__msg__State__Sequence__init(
+    &res_in->available_states,
+    sm->transition_map.states_size);
+
+  bool success = true;
+  for (unsigned int i = 0; i < sm->transition_map.states_size; ++i) {
+    res_in->available_states.data[i].id = sm->transition_map.states[i].id;
+    success &= rosidl_runtime_c__String__assign(
+      &res_in->available_states.data[i].label,
+      sm->transition_map.states[i].label);
+  }
```

Comment:
@JanStaschulat Do these lines look reasonable to you?
These are quite low-level function calls with the danger of potentially messing with memory. But I figured this would be the rcl interface to create complex msgs.

Reply:
@norro Hmm, I have no experience with the `rosidl_runtime_c__String__assign` function regarding memory management. When implementing the rclc-executor, I always used valgrind to check to memory consumption of all unit tests and examples in rclc_examples. This helped me to fix any memory issues due to dynamic memory allocation. 

Maybe @pablogs9 knows more about this rosidl function.

Reply:
Sadly, these functions do have dynamic memory calls, [check here](https://github.com/ros2/rosidl/blob/57befb87e165fc660809b30fba1a28e587263a4b/rosidl_runtime_c/src/string_functions.c#L92)

It should not be a problem if you handle `sm->transition_map.states[i].label` correctly. But when running micro-ROS in MCU calls such as `realloc` are highly discouraged.

Reply:
I would leave avoiding dynamic memory calls completely to a separate issue/PR and go forward with this version now.

Reply:
Now with pre-initialized messages, as done in rclc_parameter